### ISSUE TITLE
[bug] bactopia prepare produces empty file

### ISSUE THREAD

**Issue by lfenske-93** [bactopia/bactopia#314]
Hi 😄  

since I have a very large number of input files, I would like to have the FOFN created directly by Bactopia.

`bactopia prepare /vol/bakrep/data/Assemblies/ - "fa.gz"  > fastqs000.txt`

However, when I apply the command like this, I only get an empty file without any other error messages. 😅 
Unfortunately I could not find the problem so far, even with other input folders I get the same result.

Greetings,
Linda

**Comment by rpetit3** [comment]
Hi @lfenske-93 !

Maybe its just a typo above, but I think `--assembly_pattern` is missing.

What does this give you?
```
bactopia prepare /vol/bakrep/data/Assemblies/ --assembly_pattern fa.gz" | head
```

Here's an example:
```
ls fasta/ | grep "gz"
my-assembly-1.fa.gz
my-assembly-10.fa.gz
my-assembly-2.fa.gz
my-assembly-3.fa.gz
my-assembly-4.fa.gz
my-assembly-5.fa.gz
my-assembly-6.fa.gz
my-assembly-7.fa.gz
my-assembly-8.fa.gz
my-assembly-9.fa.gz

bactopia prepare fasta/ --assembly_pattern "*.fa.gz"
sample  runtype r1      r2      extra
my-assembly-1.fa.gz     assembly                        /tmp/fasta/my-assembly-1.fa.gz
my-assembly-10.fa.gz    assembly                        /tmp/fasta/my-assembly-10.fa.gz
my-assembly-2.fa.gz     assembly                        /tmp/fasta/my-assembly-2.fa.gz
my-assembly-3.fa.gz     assembly                        /tmp/fasta/my-assembly-3.fa.gz
my-assembly-4.fa.gz     assembly                        /tmp/fasta/my-assembly-4.fa.gz
my-assembly-5.fa.gz     assembly                        /tmp/fasta/my-assembly-5.fa.gz
my-assembly-6.fa.gz     assembly                        /tmp/fasta/my-assembly-6.fa.gz
my-assembly-7.fa.gz     assembly                        /tmp/fasta/my-assembly-7.fa.gz
my-assembly-8.fa.gz     assembly                        /tmp/fasta/my-assembly-8.fa.gz
my-assembly-9.fa.gz     assembly                        /tmp/fasta/my-assembly-9.fa.gz
```

If they are in different directories, you might have to use the `--recursive` parameter.

```
tree fastas
fastas
├── 1
│   └── my-assembly-1.fa.gz
├── 10
│   └── my-assembly-10.fa.gz
├── 2
│   └── my-assembly-2.fa.gz
├── 3
│   └── my-assembly-3.fa.gz
├── 4
│   └── my-assembly-4.fa.gz
├── 5
│   └── my-assembly-5.fa.gz
├── 6
│   └── my-assembly-6.fa.gz
├── 7
│   └── my-assembly-7.fa.gz
├── 8
│   └── my-assembly-8.fa.gz
└── 9
    └── my-assembly-9.fa.gz
    
# without --recursive its empty
bactopia prepare fastas/ --assembly_pattern "*.fa.gz"

# Now with recursive
bactopia prepare fastas/ --assembly_pattern "*.fa.gz" --recursive
sample  runtype r1      r2      extra
my-assembly-1.fa.gz     assembly                        /tmp/fastas/1/my-assembly-1.fa.gz
my-assembly-10.fa.gz    assembly                        /tmp/fastas/10/my-assembly-10.fa.gz
my-assembly-2.fa.gz     assembly                        /tmp/fastas/2/my-assembly-2.fa.gz
my-assembly-3.fa.gz     assembly                        /tmp/fastas/3/my-assembly-3.fa.gz
my-assembly-4.fa.gz     assembly                        /tmp/fastas/4/my-assembly-4.fa.gz
my-assembly-5.fa.gz     assembly                        /tmp/fastas/5/my-assembly-5.fa.gz
my-assembly-6.fa.gz     assembly                        /tmp/fastas/6/my-assembly-6.fa.gz
my-assembly-7.fa.gz     assembly                        /tmp/fastas/7/my-assembly-7.fa.gz
my-assembly-8.fa.gz     assembly                        /tmp/fastas/8/my-assembly-8.fa.gz
my-assembly-9.fa.gz     assembly                        /tmp/fastas/9/my-assembly-9.fa.gz
```

Let me know if this helps! And if you have any feedback on how to improve the prepare command let me know. 

One thing that stands out to me is, it should be more explicit to the user when nothing is found.

Cheers,
Robert

**Comment by lfenske-93** [comment]
Hi,

yeah, there was a typo in my comment, I tried with:
`bactopia prepare /vol/bakrep/data/Assemblies/ -a "fa.gz" > fastqs000.txt`

However, with the use of `--assembly_pattern` it actually worked. 
I thought the correct parameter would have been `--assembly_ext` or `-a`, which was not perfect to understand in the usage. 😅
In any case, it would be handy if some kind of error message could be issued when no files matching the pattern could be found.

Maybe it would already help if the documentation gave a few more examples to create, like you showed me here in the issue. 🤷🏻‍♀️ 

Thanks in any case for the quick help. :)

**Comment by rpetit3** [comment]
Will do! Thank you for the feedback.

I'm also going to revisit merging `--assembly_ext` and `--assembly_pattern`. I think it might be possible to drop them. I also get confused on which one to use as well.

**Comment by rpetit3** [comment]
Made some changes to `bactopia prepare`

`--fastq_pattern` and `--assembly_pattern` are no longer needed. Just `-f,--fastq_ext` and `-a,--assembly_ext`

I've also added , an error message if nothing is found:
```
Unable to find any samples in fastas/. Please try adjusting the following parameters to fit your needs.
Values Used:
    --fastq_ext => .fastq.gz
    --fastq_separator => _
    --pe1_pattern => [Aa]|[Rr]1|1
    --pe2_pattern => [Bb]|[Rr]2|2
    --assembly_ext => .fasta.gz

You can also use '--examples' to see a few examples of using bactopia prepare
```

Finally, I plan on adding example use cases to the docs website, but sometimes its nice to see it on the command line. So I added a parameters called `--examples`. Which will print a few example use cases. (_let me know if you have one in particular you'd like to see_)

```
bactopia prepare . --examples

# Example '*_001.fastq.gz' FASTQ files:
bactopia prepare fastqs/ --fastq_ext '_001.fastq.gz' | head -n
sample  runtype r1      r2      extra
sample01        paired-end      /fastqs/sample01_R1_001.fastq.gz        /fastqs/sample01_R2_001.fastq.gz
sample02        paired-end      /fastqs/sample02_R1_001.fastq.gz        /fastqs/sample02_R2_001.fastq.gz
sample03        single-end      /fastqs/sample03_001.fastq.gz

# Example '*.fq.gz' FASTQ files:
bactopia prepare fastqs/ --fastq_ext '.fq.gz'
sample  runtype r1      r2      extra
sample01        single-end      /home/robert_petit/bactopia/fastqs/sample01.fq.gz
sample02        single-end      /home/robert_petit/bactopia/fastqs/sample02.fq.gz
sample03        single-end      /home/robert_petit/bactopia/fastqs/sample03.fq.gz

# Example "*.fasta.gz" FASTA files:
bactopia prepare fastqs/ --assembly_ext '.fasta.gz'
sample  runtype r1      r2      extra
sample01        assembly                        /home/robert_petit/bactopia/temp/fastas/sample01.fasta.gz
sample02        assembly                        /home/robert_petit/bactopia/temp/fastas/sample02.fasta.gz
sample03        assembly                        /home/robert_petit/bactopia/temp/fastas/sample03.fasta.gz

# Example changing the separator:
bactopia prepare fastqs/ --fastq_separator '.'
sample  runtype r1      r2      extra
my_sample01     ont     /home/robert_petit/bactopia/temp/fastqs/my_sample01.fastq.gz
my_sample02     ont     /home/robert_petit/bactopia/temp/fastqs/my_sample02.fastq.gz
my_sample03     ont     /home/robert_petit/bactopia/temp/fastqs/my_sample03.fastq.gz
```

**Comment by lfenske-93** [comment]
Yes, the examples are definitely very helpful! 😊 
If I think of another example, I'll definitely let you know.

Is the sample name then also adapted directly? In the previous version, this always consists of the complete name with file extension.

But I'm also not sure yet if I understand what exactly ` --fastq_separator` does. 😅

**Comment by creamerk** [comment]
Hi @rpetit3 -- related to this question of preparing & checking Fofns, has the  "--check_fastqs" been depreciated in v2.0.3? I was trying to run it (ie `bactopia --check_fastqs fofn.txt` and it was giving the error: 
```
WARN: Found unexpected parameters:
* --check_fastqs: fofn.txt
- Ignore this warning: params.schema_ignore_params = "check_fastqs" 
```
Thank you!

**Comment by rpetit3** [comment]
@creamerk It didn't make it into v2, and I need to update the docs to reflect that.

If you think its worthwhile I can get it added back. 

One question I do have for both of you though. What are your thoughts on renaming `--fastqs` to something like `--samples` or `--fofn`? 

I ask because when `--fastqs` was originally created, only FASTQs could be used as inputs. But now you can include FASTQs and assemblies, so sometimes its kind of confusing to say `--fastqs assemblies.txt`.

**Comment by creamerk** [comment]
I agree, a `-samples` would make sense! And then as long as it's clear what the difference between `-sample` = name of a single sample and `-samples` = names and locations of multiple samples-- that would hopefully be clear! 

@rpetit3 re: the `--check_fastqs` , I think it might be useful to have for large datasets (and to make sure if the `prepare` function was used-- I was running ~100 genomes so I wanted to check my handmade Fofn was correct before submitting my batch job. Instead, I was just extra careful and used `realpath *R1*` etc to list the complete paths in the Fofn which I reasoned would hopefully prevent any issues of file location.

**Comment by rpetit3** [comment]
Great feedback, I've learned towards `--samples` for about a year now. So let's go with that.

Thoughts on renaming `--sample` to `--sample_name`? Or leave it as is?

I'll work on getting `--check_samples` back in there

**Comment by lfenske-93** [comment]
I think the difference between `--samples` and `--sample` should be quickly understood.
However, I could imagine that at some point there would be confusion regarding  `--accession`.
So maybe `--sample_name` and `--accession_name`?

**Comment by rpetit3** [comment]
@lfenske-93 that is a great point! I'm going to leave it at `--sample` to keep it similar to `--accession`. 

Thank you very much for bringing that up, I completely over looked it

**Comment by rpetit3** [comment]
Hi @lfenske-93 and @creamerk 

I released v2.1.0 (https://github.com/bactopia/bactopia/releases/tag/v2.1.0) which should fix includes the changes from the dev build you were using. Let me know if you have any other ideas to improve `bactopia prepare` and FOFN usage. 

If you have some time to spare, please consider filling out my Bactopia survey.  I use the feedback to direct the future developments of Bactopia. (_If you've already filled it out, thank you so much!_)

https://forms.gle/PTgUxEBwCsCdbQjr7

Cheers,
Robert
